sympy/ntheory/continued_fraction.py

```diff
@@ -58,20 +134,41 @@ def continued_fraction_periodic(p, q, d=0):
 
     """
     from sympy.core.compatibility import as_int
-    from sympy.functions import sqrt
+    from sympy.functions import sqrt, floor
+    from sympy.ntheory.primetest import is_square
+
+    p, q, d, s = list(map(as_int, [p, q, d, s]))
 
-    p, q, d = list(map(as_int, [p, q, d]))
-    sd = sqrt(d)
+    if d < 0:
+        raise ValueError("s should be a non-negative, not %s " % d)
 
     if q == 0:
-        raise ValueError("The denominator is zero.")
+        raise ValueError("The denominator should be positive.")
 
-    if d < 0:
-        raise ValueError("Delta supposed to be a non-negative "
-                         "integer, got %d" % d)
-    elif d == 0 or sd.is_integer:
-        # the number is a rational number
-        return list(continued_fraction_iterator(Rational(p + sd, q)))
+    if not s:
+        d = 0
+
+    # check for rational case
+    sd = sqrt(d) if d else 0
+    if not d or sd.is_Integer:
+        return list(continued_fraction_iterator(Rational(p + s*sd, q)))
+
+    # irrational case with sd != Integer
+    if q < 0:
+        p, q, s = -p, -q, -s
+
+    n = (p + s*sd)/q
+    if n < 0:
+        w = floor(-n)
+        f = -n - w
+        one_f = continued_fraction(1 - f)
+        if type(one_f[0]) is list:
+            one_f = [0, one_f[0]]
```

Comment:
I haven't been able to come up with an `n` that will cover this line. And if I could, I think this should be
`one_f = [one_f[0], one_f[1:]+one_f[:1]]  # e.g. [[1,2,3]] -> [1, [2, 3, 1]]`